File: frontend/src/utils/sections.js
// sections.js — clasificación de posiciones en SECCIONES de Renta Fija.
// Espejo de backend/importing/sections.py: misma lógica para que front y back
// coincidan. Una sección = (categoría, moneda): Bonos/Letras/FCI × USD/ARS.
// Renta variable (acciones/CEDEARs/cripto) → null.
import { getBondMeta } from './bondMeta'

// Renta fija conocida por catálogo: bono soberano/CER u ON corporativa. Excluye
// ETFs (TIP/GLD/etc. son renta variable aunque estén en BOND_META).
const _FIXED_META_TYPES = new Set(['sovereign', 'cer', 'corporate'])
export function isKnownArBond(symbol) {
  const m = getBondMeta(symbol)
  return !!m && _FIXED_META_TYPES.has(m.type)
}

export const CATEGORY_BONO = 'BONO'
export const CATEGORY_LETRA = 'LETRA'
export const CATEGORY_FCI = 'FCI'

const CATEGORY_LABEL = { BONO: 'Bonos', LETRA: 'Letras', FCI: 'FCI' }

// Patrón de ticker de letra/LECAP: letra inicial + día (1-2 díg) + código-mes + año
// (1 díg). Espeja maturity._LETRA_RX del backend (S28N5, T13F6, X23N3…).
const LETRA_RX = /^[A-Z]\d{1,2}[EFMAYJLGSOND]\d$/

export function isLetraTicker(symbol) {
  return LETRA_RX.test((symbol || '').trim().toUpperCase())
}

export function normCcy(currency) {
  const c = (currency || '').trim().toUpperCase()
  return (c === 'USD' || c === 'USDT') ? 'USD' : 'ARS'
}

// Devuelve {category, currency} si la posición es renta fija, o null si no.
export function positionSection(assetType, symbol, currency) {
  const at = (assetType || '').trim().toUpperCase()
  const sym = (symbol || '').trim().toUpperCase()
  const ccy = normCcy(currency)
  if (at === 'FUND') return { category: CATEGORY_FCI, currency: ccy }
  // Letra ANTES que bono (matchea el patrón aunque no esté en el catálogo).
  if (sym && isLetraTicker(sym)) return { category: CATEGORY_LETRA, currency: ccy }
  const FIXED = at === 'BOND' || at === 'BONO' || at === 'ON' || at === 'LETRA' || at === 'LECAP'
  if (FIXED || isKnownArBond(sym)) return { category: CATEGORY_BONO, currency: ccy }
  return null
}

export function isFixedIncome(p) {
  return positionSection(p.asset_type, p.asset, p.currency) != null
}

export function sectionKey(category, currency) {
  return `${category}|${normCcy(currency)}`
}

export function sectionLabel(category, currency) {
  return `${CATEGORY_LABEL[category] || category} ${normCcy(currency)}`
}

// Orden de presentación de las secciones.
const ORDER = ['BONO|USD', 'BONO|ARS', 'LETRA|USD', 'LETRA|ARS', 'FCI|USD', 'FCI|ARS']
export function sortSectionKeys(keys) {
  return [...keys].sort((a, b) => {
    const ia = ORDER.indexOf(a), ib = ORDER.indexOf(b)
    return (ia < 0 ? 99 : ia) - (ib < 0 ? 99 : ib)
  })
}

File: backend/importing/sections.py
"""Clasificación de posiciones en SECCIONES de Renta Fija (cross-broker).

Una "sección" es una agrupación de PRESENTACIÓN por (categoría, moneda):
  • Bonos USD / Bonos ARS
  • Letras USD / Letras ARS
  • FCI USD / FCI ARS

Las posiciones NO se mueven de broker (la valuación y el neteo cross-currency
dependen del broker) — esto solo las clasifica para agruparlas en la vista y para
el borrado/restore por sección. La renta variable (acciones/CEDEARs/cripto)
devuelve None (no es renta fija).
"""
from __future__ import annotations
from typing import Optional, Tuple

from .maturity import letra_maturity

_FIXED_INCOME_TYPES = {"BOND", "BONO", "ON", "LETRA", "LECAP"}

CATEGORY_BONO = "BONO"
CATEGORY_LETRA = "LETRA"
CATEGORY_FCI = "FCI"

_CATEGORY_LABEL = {CATEGORY_BONO: "Bonos", CATEGORY_LETRA: "Letras", CATEGORY_FCI: "FCI"}


def _is_known_ar_bond(symbol: str) -> bool:
    try:
        from ai.ar_bonds_metadata import is_known_ar_bond
    except Exception:
        return False
    return bool(is_known_ar_bond(symbol))


def _norm_ccy(currency: Optional[str]) -> str:
    """Moneda de la sección: USD (incluye USDT, la convención interna del sibling)
    o ARS (default)."""
    c = (currency or "").strip().upper()
    return "USD" if c in ("USD", "USDT") else "ARS"


def position_section(asset_type: Optional[str], symbol: Optional[str],
                     currency: Optional[str]) -> Optional[Tuple[str, str]]:
    """Devuelve (categoría, moneda) si la posición es renta fija, o None si no.

    categoría ∈ {BONO, LETRA, FCI}; moneda ∈ {USD, ARS}. La detección de Letra es
    por el patrón del ticker (maturity.letra_maturity), independiente del asset_type
    (IEB/otros no lo etiquetan). FCI = asset_type FUND. Bono = resto de renta fija."""
    at = (asset_type or "").strip().upper()
    sym = (symbol or "").strip().upper()
    ccy = _norm_ccy(currency)

    if at == "FUND":
        return (CATEGORY_FCI, ccy)
    # Letra ANTES que bono: un ticker de letra (S28N5, T13F6…) matchea el patrón
    # aunque no esté en el catálogo de bonos.
    if sym and letra_maturity(sym) is not None:
        return (CATEGORY_LETRA, ccy)
    if at in _FIXED_INCOME_TYPES or _is_known_ar_bond(sym):
        return (CATEGORY_BONO, ccy)
    return None


def section_key(category: str, currency: str) -> str:
    """Clave estable de sección para el dato/endpoint: 'BONO|USD'."""
    return f"{category}|{_norm_ccy(currency)}"


def section_label(category: str, currency: str) -> str:
    """Etiqueta legible: 'Bonos USD'."""
    return f"{_CATEGORY_LABEL.get(category, category)} {_norm_ccy(currency)}"


def parse_section_key(key: str) -> Optional[Tuple[str, str]]:
    """'BONO|USD' → ('BONO','USD'). None si inválida."""
    parts = (key or "").split("|")
    if len(parts) != 2:
        return None
    cat, ccy = parts[0].strip().upper(), parts[1].strip().upper()
    if cat not in (CATEGORY_BONO, CATEGORY_LETRA, CATEGORY_FCI) or ccy not in ("USD", "ARS"):
        return None
    return (cat, ccy)
